src/Dotnet.Installer.Console/Commands/ListCommand.cs

```diff
@@ -111,4 +133,181 @@ string ComponentStatus(string key, int majorVersion, bool isEol)
+            // The underlying list is not long. Therefore, the re-enumeration is cheaper than creating a new list/array.
+            if (!installedRootComponents.Any()) return;
+
+            IImmutableList<SnapInfo> installedSnaps;
+
+            try
+            {
+                installedSnaps = await _snapService.GetInstalledSnaps(cancellationToken).ConfigureAwait(false);
+            }
+            catch (ApplicationException exception)
+            {
+                anyFailure = true;
+                _logger.LogDebug(exception.Message);
+                return;
+            }
+
+            // ReSharper disable once PossibleMultipleEnumeration (see reasoning above)
+            foreach (var component in installedRootComponents)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                SnapInfo installedSnap;
+
+                try
+                {
+                    installedSnap = installedSnaps.Single(snap => snap.Name == component.Key);
+                }
+                catch (InvalidOperationException)
+                {
+                    anyFailure = true;
+                    _logger.LogDebug($"Could not find installed component {component.Key}");
+                    continue;
+                }
+
+                try
+                {
+                    componentVersions[component.Key] = installedSnap.ParseVersionAsDotnetVersion();
+                }
+                catch (ApplicationException exception)
+                {
+                    anyFailure = true;
+                    _logger.LogDebug(exception.Message);
+                }
+            }
+        }
+
+        async Task GetInstalledNonRootComponentsVersions(CancellationToken cancellationToken = default)
+        {
+            // ensures that this action will be run asynchronously
+            await Task.Yield();
+
+            foreach (var component in components.Where(c => c is { IsRoot: false, IsInstalled: true }))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    var version = component.GetDotnetVersion(_manifestService, _fileService);
+                    componentVersions[component.Key] = version;
+                }
+                catch (ApplicationException exception)
+                {
+                    anyFailure = true;
+                    _logger.LogDebug(exception.Message);
+                }
+            }
+        }
+
+        Task GetUninstalledComponentsVersions(CancellationToken cancellationToken = default)
+        {
+            return Parallel.ForEachAsync(
+                components.Where(c => !c.IsInstalled),
+                cancellationToken,
+                GetUninstalledComponentVersion);
+        }
+
+        async ValueTask GetUninstalledComponentVersion(Component component, CancellationToken cancellationToken = default)
```

Comment:
**question:** Why `ValueTask` instead of `Task` here?

Reply:
The `Parallel.ForEachAsync` method (which calls `GetUninstalledComponentVersion`; see the lines above) expects that a `ValueTask` is returned and there is no overload that allows to return `Task`. See https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync?view=net-8.0